File: iOS/TUILiveKit/Source/State/Operation/UserState.swift
//
//  UserState.swift
//  VoiceRoom
//
//  Created by aby on 2024/3/7.
//

import Foundation
import RTCRoomEngine
import ImSDK_Plus

struct UserState: Encodable {
    var selfInfo: User = User()
    var userList: [User] = []
    var myFollowingUserList: Set<User> = []
    var hasAudioStreamUserList: Set<String> = []
    var hasVideoStreamUserList: Set<String> = []
    var speakingUserList: Set<String> = []
}

struct User: Codable {
    
    var userId: String = ""
    var name: String = ""
    var avatarUrl: String = ""
    var role: TUIRole = .generalUser
    var fansCount: Int = 0
    
    init(userInfo: TUIUserInfo) {
        self.userId = userInfo.userId
        self.name = userInfo.userName
        self.avatarUrl = userInfo.avatarUrl
        self.role = userInfo.userRole
    }
    
    init(loginInfo: TUILoginUserInfo) {
        self.userId = loginInfo.userId
        self.name = loginInfo.userName
        self.avatarUrl = loginInfo.avatarUrl
    }
    
    init(userInfo: V2TIMUserFullInfo) {
        self.userId = userInfo.userID
        self.name = userInfo.nickName
        self.avatarUrl = userInfo.faceURL
    }
    
    init(userId: String) {
        self.userId = userId
    }
    
    init() {}
}

extension User: Identifiable {
    var id: String {
        return self.userId
    }
}

extension User: Hashable {
    func hash(into hasher: inout Hasher) {
        hasher.combine(userId)
        hasher.combine(name)
        hasher.combine(avatarUrl)
        hasher.combine(role)
    }
    
    static func == (lhs: User, rhs: User) -> Bool {
        return  lhs.userId == rhs.userId && lhs.name == rhs.name && lhs.avatarUrl == rhs.avatarUrl
    }
}

extension TUIRole: Codable {
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let rawValue = try container.decode(UInt.self)
        self = TUIRole(rawValue: rawValue) ?? .generalUser
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(self.rawValue)
    }
}

File: Flutter/livekit/lib/voice_room/state/user_state.dart
import 'package:flutter/cupertino.dart';
import 'package:rtc_room_engine/rtc_room_engine.dart';
import 'package:tencent_cloud_chat_sdk/models/v2_tim_user_full_info.dart';
import 'voice_define.dart';

class UserState {
  User selfInfo = User();
  ValueNotifier<List<User>> userList = ValueNotifier([]);
  ValueNotifier<Set<User>> myFollowingUserList = ValueNotifier({});
  ValueNotifier<Set<String>> hasAudioStreamUserList = ValueNotifier({});
  ValueNotifier<Set<String>> speakingUserList = ValueNotifier({});
  ValueNotifier<User> enterUser = ValueNotifier(User());
}

class User {
  String userId;
  String name;
  String avatarUrl;
  TUIRole role;
  int fansCount;
  ValueNotifier<LinkStatus> linkStatus = ValueNotifier(LinkStatus.none);

  User(
      {this.userId = '',
      this.name = '',
      this.avatarUrl = '',
      this.role = TUIRole.generalUser,
      this.fansCount = 0,
      LinkStatus linkStatus = LinkStatus.none})
      : linkStatus = ValueNotifier(linkStatus);

  User.fromTUIUserInfo(TUIUserInfo userInfo)
      : this(
            userId: userInfo.userId,
            name: userInfo.userName,
            avatarUrl: userInfo.avatarUrl,
            role: userInfo.userRole);

  User.fromTUILoginUserInfo(TUILoginUserInfo loginUserInfo)
      : this(
            userId: loginUserInfo.userId,
            name: loginUserInfo.userName ?? '',
            avatarUrl: loginUserInfo.avatarUrl ?? '');

  User.fromV2TIMUserFullInfo(V2TimUserFullInfo imUserInfo)
      : this(
            userId: imUserInfo.userID ?? '',
            name: imUserInfo.nickName ?? '',
            avatarUrl: imUserInfo.faceUrl ?? '');
}
